extensions/builder/build-pipe.ts

```diff
@@ -1,48 +1,126 @@
+import GeneralError from 'bit-bin/dist/error/general-error';
+import { ComponentMap } from '@teambit/component';
 import { Logger } from '@teambit/logger';
 import Bluebird from 'bluebird';
 import prettyTime from 'pretty-time';
-
+import { ArtifactFactory, ArtifactList } from './artifact';
+import { BuildContext, BuildTask, BuildResults } from './build-task';
 import { InvalidTask } from './exceptions';
-import { TaskProcess } from './task-process';
-import { BuildContext, BuildResults, BuildTask } from './types';
+import { ComponentResult } from './types';
+
+export type BuildPipeResults = {
+  /**
+   * results of all tasks executed in the build pipeline.
+   */
+  results: TaskResults[];
+
+  /**
+   * start time of the build pipeline.
+   */
+  startTime: number;
+
+  /**
+   * end time of the build pipeline.
+   */
+  endTime: number;
+};
+
+export type TaskResults = {
+  /**
+   * component build results.
+   */
+  componentsResults: ComponentResult[];
+
+  /**
+   * artifacts generated by the build pipeline.
+   */
+  artifacts: ComponentMap<ArtifactList>;
+
+  /**
+   * timestamp of start initiation.
+   */
+  startTime: number;
+
+  /**
+   * timestamp of task completion.
+   */
+  endTime: number;
+};
 
 export class BuildPipe {
   constructor(
     /**
      * array of services to apply on the components.
      */
     readonly tasks: BuildTask[],
-    readonly logger: Logger
+    readonly logger: Logger,
+    readonly artifactFactory: ArtifactFactory
   ) {}
 
   /**
    * execute a pipeline of build tasks.
    */
-  async execute(buildContext: BuildContext): Promise<BuildResults[]> {
+  async execute(buildContext: BuildContext): Promise<BuildPipeResults> {
+    const startTime = Date.now();
     const longProcessLogger = this.logger.createLongProcessLogger('running tasks', this.tasks.length);
     const results = await Bluebird.mapSeries(this.tasks, async (task: BuildTask) => {
       if (!task) {
         throw new InvalidTask(task);
       }
-      const taskName = `${task.extensionId} ${task.description || ''}`;
+      const taskName = `${task.id} ${task.description || ''}`;
       longProcessLogger.logProgress(taskName);
       const startTask = process.hrtime();
+      const taskStartTime = Date.now();
       const taskResult = await task.execute(buildContext);
-      const taskProcess = new TaskProcess(task, taskResult, buildContext, this.logger);
-      taskProcess.throwIfErrorsFound();
+      const endTime = Date.now();
+      // TODO: move this function to the upper scope. that should happen from the consumer of the service.
```

Comment:
@ranm8 , the problem with having this inside the onTag, is that I have to pass all the errors props from the different levels, into the onTag.